src/OpenDiffix.Core/Analyzer.fs

```diff
@@ -419,13 +424,30 @@ let rec private normalizeBucketLabelExpression expression =
     FunctionExpr(ScalarFunction fn, List.map normalizeBucketLabelExpression args @ extraArgs)
   | _ -> expression
 
+let private untrustedAllowsRange arg =
+  match arg with
+  // "money-style" numbers, i.e. 1, 2, or 5 preceeded by or followed by zeros: ⟨... 0.1, 0.2, 0.5, 1, 2, 5, 10, 20, ...⟩
+  | Constant (Real c) -> Regex("^(1|2|5)\.0+e(\-|\+)[0-9]+$").Match(sprintf "%.15e" c).Success
+  | Constant (Integer c) -> Regex("^(1|2|5)0*$").Match(sprintf "%i" c).Success
+  | _ -> false
+
+let private validateBucketLabelExpression accessLevel expression =
+  if accessLevel = PublishUntrusted then
+    match expression with
+    | FunctionExpr (ScalarFunction FloorBy, [ _; arg ]) when untrustedAllowsRange arg -> ()
+    | FunctionExpr (ScalarFunction Substring, [ _; fromArg; _ ]) when fromArg = (1L |> Integer |> Constant) -> ()
+    | _ -> failwith "Generalization used in the query is not allowed in untrusted analyst mode"
+
+  expression
+
 let private computeNoiseLayers anonParams query =
   let rangeColumns = collectRangeColumns anonParams query.From
 
   let sqlSeed =
     query.GroupBy
     |> Seq.map (
       normalizeBucketLabelExpression
+      >> validateBucketLabelExpression anonParams.AccessLevel
```

Comment:
Ideally, this should be under `QueryValidator`, just like other validations that the seed calculation is currently doing. I'd like to proceed as is and consider a larger refactoring in a future PR.